proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1179 @@
+
+Alongside this work, it may make sense to add a `Span` alternative to `withContiguousStorageIfAvailable()`, `RawSpan` alternative to `withUnsafeBytes`, etc., and seek to deprecate any closure-based API around unsafe pointers.
+
+##### Sharing piecewise-contiguous memory
+
+Some types store their internal representation in a piecewise-contiguous manner, such as trees and ropes. Some operations naturally return information in a piecewise-contiguous manner, such as network operations. These could supply results by iterating through a list of contiguous chunks of memory.
+
+##### Safe mutations of memory with `MutableSpan<T>`
+
+Some data structures can delegate mutations of their owned memory. In the standard library we have `withMutableBufferPointer()`, for example.
+
+The `UnsafeMutableBufferPointer` passed to a `withUnsafeMutableXXX` closure-style API is unsafe in multiple ways:
+
+1. The pointer itself is unsafe and unmanaged
+2. `subscript` is only bounds-checked in debug builds of client code
+3. It might escape the duration of the closure
+4. Exclusivity of writes is not enforced
+5. Initialization of any particular memory address is not ensured
+
+I.e., it is unsafe in all the ways `UnsafeBufferPointer`-passing closure APIs are unsafe in addition to being unsafe in exclusivity and in initialization.
+
+Loading an uninitialized non-`BitwiseCopyable` value leads to undefined behavior. Loading an uninitialized `BitwiseCopyable` value does not immediately lead to undefined behavior, but it produces a garbage value which may lead to misbehavior of the program.
+
+A `MutableSpan<T>` should provide a better, safer alternative to mutable memory in the same way that `Span<T>` provides a better, safer read-only type. `MutableSpan<T>` would also automatically enforce exclusivity of writes.
+
+However, it alone does not track initialization state of each address, and that will continue to be the responsibility of the developer.
+
+##### Delegating initialization of memory with `OutputSpan<T>`
+
+Some data structures can delegate initialization of their initial memory representation, and in some cases the initialization of additional memory. In the standard library we have `Array.init(unsafeUninitializedCapacity:initializingWith:)` and `String.init(unsafeUninitializedCapacity:initializingUTF8With:)`. A safer abstraction for initialization would make such initializers less dangerous, and would allow for a greater variety of them.
+
+`OutputSpan<T>` would need run-time bookkeeping (e.g. a bitvector with a bit per-address) to track initialization state to safely support random access and random-order initialization.
+
+Alternatively, a divide-and-conqueor style initialization order might be solvable via an API layer without run-time bookkeeping, but with more complex ergonomics.
```

Comment:
The primary use case of `OutputSpan` is to serve as the unit of data exchange in consuming iteration. (E.g., to allow containers of noncopyable types to efficiently implement an `append` operation that takes a generic consuming sequence.)

Allowing divide-and-conquer initialization is a second order benefit, at best. Luckily, that doesn't need bitmaps, either.

Reply:
improved the description of `OutputSpan`.